GG/src/Font.cpp

```diff
@@ -478,6 +484,186 @@ std::pair<std::size_t, CPSize> GG::LinePositionOf(CPSize index, const std::vecto
+                // matching close tags, or updating the stack when
+                // matching open tags
+                m_EVERYTHING =
+                    ('<' >> OPEN_TAG_NAME >> xpr::repeat<0, 9>(+xpr::blank >> TAG_PARAM) >> '>') |
+                    ("</" >> OPEN_TAG_NAME >> '>') |
+                    (whitespace_tag = WHITESPACE) |
+                    (text_tag = TEXT);
+            }
+        }
+
+        xpr::sregex& BindRegexToText(const std::string& new_text, bool ignore_tags) {
+            if (!m_tag_stack.empty()) {
+                std::stack<Font::Substring> empty_stack;
+                std::swap(m_tag_stack, empty_stack);
+            }
+            m_text = &new_text;
+            m_ignore_tags = ignore_tags;
+            return m_EVERYTHING;
+        }
+
+    private:
+
+        bool MatchesKnownTag(const boost::xpressive::ssub_match& sub) {
+            return m_ignore_tags ? false : m_known_tags->count(sub.str()) != 0;
+        }
+
+        bool MatchesTopOfStack(const boost::xpressive::ssub_match& sub) {
+            bool retval = m_tag_stack.empty() ? false : m_tag_stack.top() == sub;
+            if (retval) {
+                m_tag_stack.pop();
+                if (m_tag_stack.empty() || m_tag_stack.top() != PRE_TAG)
+                    m_ignore_tags = false;
+            }
+            return retval;
+        }
+
+        const std::string* m_text;
+        const boost::unordered_set<std::string>* m_known_tags;
+        bool m_ignore_tags;
+
+        // m_tag_stack is used to track XML opening/closing tags.
+        std::stack<Font::Substring> m_tag_stack;
+
+        // The combined regular expression.
+        xpr::sregex m_EVERYTHING;
+    };
+
+    /** TagHandler stores a set of all known tags and provides pre-compiled regexs for those tags.
+
+     Known tags are tags that will be parsed into TextElement OPEN_TAG or CLOSE_TAG. */
+    class TagHandler {
+    public:
+        TagHandler() :
+            m_known_tags(),
+            m_regex_w_tags(m_known_tags, false),
+            m_regex_w_tags_skipping_unmatched(m_known_tags, true)
+        {}
+
+        /** Add a tag to the set of known tags.*/
+        void Insert(const std::string& tag) {
+            m_known_tags.insert(tag);
+        }
+
+        /** Remove a tag from the set of known tags.*/
+        void Erase(const std::string& tag) {
+            m_known_tags.erase(tag);
+        }
+
+        /** Remove all tags from the set of known tags.*/
+        void Clear() {
+            m_known_tags.clear();
+        }
+
+        bool IsKnown(const std::string &tag) {
+            return m_known_tags.find(tag) != m_known_tags.end();
+        }
+
+        // Return a regex bound to \p text using the currently known
+        // tags.  If required \p ignore_tags and/or \p strip_unpaired_tags.
+        xpr::sregex & Regex(const std::string& text, bool ignore_tags, bool strip_unpaired_tags = false) {
```

Comment:
Fixed the space.

It can't be const because the `BindRegexToText` clears and returns either `m_regex_w_tags` or `m_regex_w_tags_skipping_unmatched`, neither of which is mutable.